crates/lib/src/execution_data.rs

```diff
@@ -1,60 +1,114 @@
+use ndarray::prelude::*;
+
+use qcs_api_client_grpc::models::controller::{readout_values, ReadoutValues};
 
 use crate::RegisterData;
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// The result of executing an [`Executable`](crate::Executable)
 #[derive(Debug, Clone, PartialEq)]
-pub struct Qvm {
+pub struct ExecutionData {
     /// The readout data that was read from the [`Executable`](crate::Executable).
     /// Key is the name of the register, value is the data of the register after execution.
-    pub registers: HashMap<Box<str>, RegisterData>,
+    pub readout_data: ReadoutMap,
     /// The time it took to execute the program on the QPU, not including any network or queueing
     /// time. If paying for on-demand execution, this is the amount you will be billed for.
     ///
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
 
-/// A mapping of readout fields to their [`ReadoutValues`].
+/// An enum representing every possible readout type
+#[derive(Clone, Copy, Debug, EnumAsInner, PartialEq)]
+pub enum ReadoutValue {
+    /// An integer readout value
+    Integer(i32),
+    /// A real numbered readout value
+    Real(f64),
+    /// A complex numbered readout value
+    Complex(Complex64),
+}
+
+/// A matrix where rows are the values of a memory index across all shots, and columns are values
+/// for all memory indexes of a given shot.
+pub type RegisterMatrix = Array2<Option<ReadoutValue>>;
+
+/// A mapping of readout fields to their [`ReadoutValue`]s.
 #[derive(Debug, Clone, PartialEq)]
 #[repr(transparent)]
-pub struct ReadoutMap(HashMap<MemoryReference, ReadoutValues>);
+pub struct ReadoutMap(HashMap<String, RegisterMatrix>);
 
 impl ReadoutMap {
-    /// Given a known readout field name and index, return the result's [`ReadoutValues`], if any.
+    /// Returns a [`ReadoutValue`] for the given memory index and shot number, if any
     #[must_use]
-    pub fn get_readout_values(&self, field: String, index: u64) -> Option<ReadoutValues> {
-        let readout_values = self.0.get(&MemoryReference { name: field, index })?;
+    pub fn get_value(
+        &self,
+        register_name: &str,
+        index: usize,
+        shot_num: usize,
+    ) -> Option<ReadoutValue> {
+        self.0
+            .get(register_name)
+            .and_then(|matrix| matrix.get((index, shot_num)))
+            .copied()?
+    }
 
-        Some(readout_values.clone())
+    /// Returns a vector of the [`ReadoutValue`]s in the given register at a particular memory
+    /// index across all shots.
+    #[must_use]
+    pub fn get_values_by_memory_index(
+        &self,
+        register_name: &str,
+        index: usize,
+    ) -> Option<Vec<Option<ReadoutValue>>> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if index >= matrix.nrows() {
```

Comment:
Why isn't this an error?

Reply:
I suppose it could be, and I'm open to that idea. I went with this method because it aligns with the default behavior of standard library containers like `Vec`. `Vec<T>.get(i)` returns `Option<T>` and returns `None` if `i` is out of bounds.